[OSD-26117] Add command to start ssm session for nodes of HCP clusters

Dee-6777 opened the pull request:
### What type of PR is this?

_feature_

### What this PR does / Why we need it?
This PR will add a subcommand named "ssm" to start a ssm session for HCP with a given node name in it's management cluster
Example command: `ocm backplane cloud ssm --node ${NODE_NAME}`

Example output:
```
ocm-backplane cloud ssm --node ip-10-0-1-99.us-west-2.compute.internal

Starting session with SessionId: RH-SRE-dmaharan.osd-h2nxx4ltqjq244fibiyvf2blxe
sh-5.1$ ls
 VGAuthService                  info                       pydoc3                           toolbox
'['                             infocmp                    pydoc3.9                         top
 afterburn                      infotocap                  python                           touch
 alias                          install                    python3                          tpm2
 amazon-ssm-agent               ionice                     python3.9                        tpm2_activatecredential
 arch                           ipcalc                     pzstd                            tpm2_certify
 arpaname                       ipcmk                      qemu-ga                          tpm2_certifyX509certutil
...

sh-5.1$ exit
exit

Exiting session with sessionId: RH-SRE-dmaharan.osd-h2nxx4ltqjq244fibiyvf2blxe.
 ```

### Which Jira/Github issue(s) does this PR fix?

_Resolves_ [OSD-26117](https://issues.redhat.com/browse/OSD-26117)

### Special notes for your reviewer

### Unit Test Coverage
#### Guidelines
- If it's a new sub-command or new function to an existing sub-command, please cover at least 50% of the code
- If it's a bug fix for an existing sub-command, please cover 70% of the code 
 
#### Test coverage checks  
- [X] Added unit tests
- [ ] Created jira card to add unit test 
- [ ] This PR may not need unit tests

### Pre-checks (if applicable)
- [X] Ran unit tests locally
- [X] Validated the changes in a cluster
- [X] Included documentation changes with PR

feichashao commented:
Follow up card to add test cases https://issues.redhat.com/browse/OSD-27393

feichashao commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,150 @@
+	"github.com/spf13/cobra"
+)
+
+var Node string
```
If this variable is not needed outside this package, we can use a lower case `node`.  https://go.dev/ref/spec#Exported_identifiers   

Though this is the only flag we need in this package, I would suggest keeping the same convention as other packages. Example:
https://github.com/openshift/backplane-cli/blob/fd3a95700cff58ecbe94bf9e921283adc08a7b0f/cmd/ocm-backplane/cloud/console.go#L58
We can define
```
var ssmArgs struct {
	node string
}
```

feichashao commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,150 @@
+// getInstanceID fetches the instance id of the given node
+func getInstanceID(Node string) (string, error) {
+	logger.Infof("Fetching instance ID for node: %s", Node)
+	command := fmt.Sprintf("oc get node %s -o jsonpath='{.spec.providerID}' | awk -F/ '{print $NF}'", Node)
```
Let's try using `client-go` whenever possible instead of calling the `oc` command directly.
Could you have a try to see if `client-go` can work for this?
https://pkg.go.dev/k8s.io/client-go/kubernetes#Clientset.CoreV1
https://pkg.go.dev/k8s.io/client-go@v0.32.0/kubernetes/typed/core/v1#CoreV1Client.Nodes

Examples:
https://github.com/openshift/backplane-cli/blob/fd3a95700cff58ecbe94bf9e921283adc08a7b0f/cmd/ocm-backplane/console/console.go#L760

Dee-6777 replied:
I can try to use this

bmeng replied:
+1 please try not execute shell commands directly within golang

feichashao commented:
@Dee-6777 Thank you for the well-struct code and a detailed README!
I left some comments, could you check them?

devppratik commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	// Construct the AWS SSM command
+	command := fmt.Sprintf("aws ssm start-session --target %s", instanceID)
+	logger.Infof("Executing command: %s", command)
+
+	cmdExec := exec.Command("sh", "-c", command)
+	cmdExec.Stdout = os.Stdout
+	cmdExec.Stderr = os.Stderr
+
+	if err := cmdExec.Run(); err != nil {
+		logger.Errorf("Failed to start SSM session: %v", err)
+		fmt.Printf("To manually start the session, run: %s\n", command)
+		return err
+	}
```
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ssm

Dee-6777 replied:
Thanks pratik, I can try to implement this

Dee-6777 replied:
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ssm - those have all the required APIs but we also need to implement the ssm plugin

bmeng replied:
Are you going to update this with aws-sdk-go? What do you mean the plugin here?

devppratik replied:
We would need not to implement plugin here. The SDK would already have the API interface to call the plugin

bmeng commented on README.md:
```diff
@@ -82,6 +82,7 @@ To setup the PS1(prompt) for bash/zsh, please follow [these instructions](https:
 | `ocm backplane upgrade`                                                     | Upgrade backplane-cli to the latest version                                              |
 | `ocm backplane version`                                                     | Display the installed backplane-cli version                                              |
 | `ocm backplane healthcheck`                                                 | Check the VPN and Proxy connectivity on the host network when experiencing isssues accessing the backplane API|
+| `ocm backplane cloud ssm --node <node-name>`                                | Start an aws ssm session for an HCP cluster                                    |
```
Please move this to close the other ocm backplane cloud commands

joshbranham commented:
> Follow up card to add test cases https://issues.redhat.com/browse/OSD-27393

Can we include tests from the get go? Filing a ticket to add tests later feels _wrong_ and likely to be forgotten.

samanthajayasinghe commented:
Thanks for the great PR .. any possibilities of adding unit tests for `cloud ssm` sub command ?

samanthajayasinghe commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	// Construct the AWS SSM command
+	command := fmt.Sprintf("aws ssm start-session --target %s", instanceID)
```
Hope you can use AWS SDK to execute ssm and it may help you to add the unit tests with fake AWS clients too.. 
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ssm

Dee-6777 replied:
Thanks Sam! I'm trying to implement this

Dee-6777 commented:
> Thanks for the great PR .. any possibilities of adding unit tests for `cloud ssm` sub command ?

Hey @samanthajayasinghe I was thinking of implementing this in a different PR since for this card [OSD-26117](https://issues.redhat.com/browse/OSD-26117) DC is to add the subcommand only. Wdyt?
Also, I've created another card [OSD-27393](https://issues.redhat.com/browse/OSD-27393) for adding the test cases.
Let me know your thoughts.

Dee-6777 commented:
> Are you going to update this with aws-sdk-go? What do you mean by the plugin here?
Hey @bmeng, Yes I'm going to update the aws-sdk-go. And the plugin here refers to the ssm plugin which is present in the SDK api interface.

samanthajayasinghe commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,163 @@
+	return awsCreds, nil
+}
+
+func getInstanceID(node string, config *rest.Config) (string, error) {
+	logger.Infof("Fetching instance ID for node: %s", node)
+
+	clientset, err := createClientSet(config)
+	if err != nil {
+		return "", fmt.Errorf("failed to create Kubernetes client: %w", err)
+	}
+
+	nodeDetails, err := clientset.CoreV1().Nodes().Get(context.TODO(), node, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get node %s: %w", node, err)
+	}
+
+	providerID := nodeDetails.Spec.ProviderID
+	if providerID == "" {
+		return "", fmt.Errorf("providerID is not set for node %s", node)
+	}
+
+	parts := strings.Split(providerID, "/")
+	instanceID := parts[len(parts)-1]
+
+	return instanceID, nil
+}
```
Any chance to move this function to /pkg/cloud/ssm module and add some tests ? 
I hope this can be tested with mocking AWS client

Dee-6777 replied:
Yes this can be tested. I'll add the test cases today.

xiaoyu74 commented:
Just as we discussed [here](https://redhat-internal.slack.com/archives/C0326L38PEH/p1738276226294129?thread_ts=1736497731.162349&cid=C0326L38PEH), better to use the existing `aws ssm sdk`  to implement, which can reduce dependencies and better improve maintainability - for example, it can reduce the need for the aws cli binary and plugins to be installed (either install on the localhost or bundled with bp-cli, which might not be a good practice)

 Also, better to have unit test included in this PR.

xiaoyu74 commented:
/hold

xiaoyu74 commented:
I've implemented the ssm.go with ssm sdk and invoking the session-manager-plugin to start/attach the ssm session and running all good. 

Probably I can create a continuation PR based on your branch, so you can check.

Dee-6777 commented:
- Implemented the ssm command using session manager plugin
- The code works after fixing parameters that were sent to the session-manager-plugin command
```
cmdArgs := []string{"session-manager-plugin", string(sessionJSON), creds.Region, "StartSession"}	
pluginCmd := exec.Command(cmdArgs[0], cmdArgs[1:]...) 
```
- Added the test cases.

xiaoyu74 commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,219 @@
+
+	logger.Infof("Starting SSM session for node: %s with Instance ID: %s", ssmArgs.node, instanceID)
+
+	cfg, err := awsConfig.LoadDefaultConfig(context.TODO(), awsConfig.WithRegion("us-west-2"))
```
Why the region is hard coded region as `us-west-2`?  Shouldn't it come from the AWS credentials we fetched as env variable?

Dee-6777 replied:
Good catch!! I added this while testing since I was getting some credentials error. Thought fetching region through creds is not working. Updating this one.

xiaoyu74 commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,219 @@
+	cf := genericclioptions.NewConfigFlags(true)
+	config, err := cf.ToRESTConfig()
+	if err != nil {
+		return nil, fmt.Errorf("--node flag is required")
```
The function `getCurrentKubeconfig()` is to load the `kubeConfig`, why the error is returned as  `--node flag is required` ?

Dee-6777 replied:
No I might have mistakenly added this in the beggining.

samanthajayasinghe commented on cmd/ocm-backplane/cloud/ssm.go:
```diff
@@ -0,0 +1,219 @@
+package cloud
```
Can you pls move ssm logic to `/pkg/cloud/ssm` and the unit tests too 
you can see existing logic/cmd separation here 
https://github.com/openshift/backplane-cli/tree/main/pkg/elevate

Dee-6777 replied:
Will sure do in this way

Dee-6777 replied:
It's a bit complicated to refactor this at the moment since there are many dependencies on other cloud package functions. Merging this pr without the refactor for now.

samanthajayasinghe replied:
All good for now , pls create a follow-up card for moving the logic and tests in the standard way.

samanthajayasinghe commented on cmd/ocm-backplane/cloud/ssm_test.go:
```diff
@@ -0,0 +1,255 @@
+	"github.com/openshift/backplane-cli/pkg/info"
+	"github.com/openshift/backplane-cli/pkg/ocm"
+	ocmMock "github.com/openshift/backplane-cli/pkg/ocm/mocks"
+	"github.com/openshift/backplane-cli/pkg/ssm/mocks"
```
You may need to update mockgen target with SSM mocks 
https://github.com/openshift/backplane-cli/blob/main/Makefile#L109

Dee-6777 replied:
Will add this

xiaoyu74 commented:
/label tide/merge-method-squash

samanthajayasinghe commented:
/lgtm
